### ISSUE TITLE
Auto add a semicolon at the end

### ISSUE THREAD

**Issue by Hungpro3112** [clangd/clangd#842]
I am so lazy and i want `clangd` auto add the semicolon at the end for me

I use `coc.clangd `for vim8.2  and really love it. Hope `clangd` can add a semicolon :)))

**Comment by HighCommander4** [comment]
A way this could work is:

  1. clang provides a diagnostic about the missing semicolon, with a fix-it that adds the semicolon
  2. the editor automatically applies the fix-it

There is a question of how does the editor know to automatically apply the fix-it. Perhaps there is an option to auto-apply all fix-its where a diagnostic has exactly one fix-it, one perhaps there is an annotation on fix-its which can be auto-applied.

(1) is an upstream clang feature that already works in some contexts and (probably) not others. If there are contexts where you do not currently get a fix-it to add the semicolon but would like to, you'll need to file a [clang bug](https://bugs.llvm.org/) to get that fixed in clang.

(2) is mostly an editor feature (with some possibility for the involvement fo LSP / clangd depending on how the auto-application works).

**Comment by kirillbobyrev** [comment]
Is this something you want to do for every line you finish and then press the Enter key? If so, then maybe just Vim's `inoremap <Enter> ;<Enter>`?

**Comment by HighCommander4** [comment]
There are many lines that don't end in semicolons (e.g. the first lines of if statements, loops, function bodies, class bodies; blank and comment lines; ...)

**Comment by kirillbobyrev** [comment]
> There are many lines that don't end in semicolons (e.g. the first lines of if statements, loops, function bodies, class bodies; blank and comment lines; preprocessor statements; ...)

You're right, I thought there are two options:

- Bind it to something other than `<Enter>` and gain some muscle memory.
- Do something more sophisticated and implement heuristics. I think there are reasonable heuristics for deciding whether the line should end with a semicolon or not (given the certain style), e.g. check for `//` or `/*` in the beginning of the line, see if the line ends with `{}<>...`

But the heuristics wouldn't work so well now that I think of some code that would break them. I initially thought it's quite easy to narrow it down to some reasonable baseline but I guess it'd be complicated in the end.

**Comment by HighCommander4** [comment]
Yeah, I briefly thought about whether `clang-format` could insert the semicolons based on its token-level model of the code, but I think such an approach would come with a risk of breakage (see also [active discussion](https://lists.llvm.org/pipermail/llvm-dev/2021-August/152070.html) about whether clang-format should be allowed to make non-whitespace changes).
